### PR TITLE
feat: byod generator

### PR DESCRIPTION
**Pull request by vgkowski**
Issue #, if available: #388

Description of changes: add a CustomDataset construct so you can BatchReplay any dataset and bring your own dataset in ARA. Also add a SynchronousGlueJob (used in the CustomDataset) to create and run a Glue PySpark job during deployment time.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: core/THIRD-PARTY-LICENSES [hunk 1]
```diff
@@ -1,75 +0,0 @@
...
-The above copyright notice and this permission notice shall be included in all
-copies or substantial portions of the Software.
-
-THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-SOFTWARE.
```
Comment (lmouhib) [inline core/THIRD-PARTY-LICENSES:75]: we need to keep this

File: core/src/data-generator/custom-dataset.ts [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    // the SSM parameter used to stored the output of the Glue job
+    const ssm = new StringParameter(this, 'JobOutputParameter', {
+      stringValue: 'minDatetime',    
+    });
+    ssm.grantWrite(glueRole);
+
+    glueRole.addManagedPolicy(ManagedPolicy.fromAwsManagedPolicyName('service-role/AWSGlueServiceRole'));
+
+    // create a KMS key for the Glue security configureation
+    const glueKey = new Key(this, 'GlueKey', {
```
Comment (lmouhib) [inline core/src/data-generator/custom-dataset.ts:127]: Can we use the KMS key available to the whole stack?

File: core/src/data-generator/custom-dataset.ts [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+        '--s3_output_path': `s3://${outputBucket.bucketName}/${props.s3Location.objectKey}`,
+        '--input_format': props.inputFormat,
+        '--datetime_column': props.datetimeColumn,
+        '--partition_range': props.partitionRange.toMinutes().toString(),
+        '--ssm_parameter': ssm.parameterName,
+        '--enable-auto-scaling': 'true',
+      },
+      role: glueRole,
+      workerCount: props.approximateDataSize ? workerNum : 100,
+      workerType: WorkerType.G_1X,
```
Comment (lmouhib) [inline core/src/data-generator/custom-dataset.ts:187]: Would it be possible to use the FLEX price type or the smaller worker recently launched?

File: core/src/data-generator/prepared-dataset.ts [hunk 4]
```diff
@@ -86,6 +94,9 @@ export interface PreparedDatasetProps {
  * 16000000600  , s3://<path>/<to>/<folder>/time_range_start=16000000600/file2.csv
  *
  * 16000000600  , s3://<path>/<to>/<folder>/time_range_start=16000000600/whichever-file....csv
+ *
```
Comment (lmouhib) [inline core/src/data-generator/prepared-dataset.ts:97]: What happen if pass both of them? Should we check if both are passed?

File: core/src/data-generator/custom-dataset.ts [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+      // add inline policy to encrypt logs
+      inlinePolicies: {
+        SecurityConfig: new PolicyDocument({
+          statements: [
+            new PolicyStatement({
+              actions: [
+                "logs:AssociateKmsKey"
+              ],
+              resources: [
+                "arn:aws:logs:us-east-1:668876353122:log-group:/aws-glue/jobs/*"
```
Comment (lmouhib) [inline core/src/data-generator/custom-dataset.ts:106]: any reason why this log group is hard coded with the account id?

File: core/src/data-generator/custom-dataset.ts [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+  JSON = 'json',
+}
+
+/**
+ * A CustomDataset is a dataset that you need to prepare for the [BatchReplayer](@link BatchReplayer) to generate data.
+ * The dataset is transformed into a [PreparedDataset](@link PreparedDataset) by a Glue Job that runs synchronously during the CDK deploy.
+ * The Glue job is sized based on the approximate size of the input data or uses autoscaling (max 100) if no data size is provided.
+ * 
+ * The Glue job is applying the following transformations to the input dataset:
+ * 1. Read the input dataset based on its format. Currently, it supports data in CSV, JSON and Parquet
```
Comment (lmouhib) [inline core/src/data-generator/custom-dataset.ts:61]: Can we provide more clarity on the date format as this is something essential for transforming and replaying the data set?

File: core/src/data-generator/custom-dataset.ts [hunk 7]
```diff
@@ -0,0 +1,228 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (lmouhib) [inline core/src/data-generator/custom-dataset.ts:1]: Can we provide as output the location of logs used by glue so a user can check potential issues/error with the job.

**Review by lmouhib — COMMENTED** [review]
Few things to discuss/address.

File: core/src/data-generator/prepared-dataset.ts (hunk 4 above)
Comment (vgkowski) [reply core/src/data-generator/prepared-dataset.ts:97]: now throws an error if both are passed

File: core/src/data-generator/custom-dataset.ts (hunk 6 above)
Comment (vgkowski) [reply core/src/data-generator/custom-dataset.ts:61]: added in construct doc

File: core/src/data-generator/custom-dataset.ts (hunk 7 above)
Comment (vgkowski) [reply core/src/data-generator/custom-dataset.ts:1]: stored as an object parameter and providing example to create a CFN output from it

**Review by lmouhib — DISMISSED** [review]
GTG, one last comment add a link to a sample/example file provided as input for data generator.

**Review by lmouhib — APPROVED** [review]

**Pull request merged**